src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -681,42 +711,56 @@ BuildShardPlacementList(ShardInterval *shardInterval)
 {
 	int64 shardId = shardInterval->shardId;
 	List *shardPlacementList = NIL;
-	Relation pgShardPlacement = NULL;
+	Relation pgPlacement = NULL;
 	SysScanDesc scanDescriptor = NULL;
 	ScanKeyData scanKey[1];
 	int scanKeyCount = 1;
 	bool indexOK = true;
 	HeapTuple heapTuple = NULL;
 
-	pgShardPlacement = heap_open(DistShardPlacementRelationId(), AccessShareLock);
+	pgPlacement = heap_open(DistPlacementRelationId(), AccessShareLock);
 
-	ScanKeyInit(&scanKey[0], Anum_pg_dist_shard_placement_shardid,
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_placement_shardid,
 				BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(shardId));
 
-	scanDescriptor = systable_beginscan(pgShardPlacement,
-										DistShardPlacementShardidIndexId(), indexOK,
+	scanDescriptor = systable_beginscan(pgPlacement,
+										DistPlacementShardidIndexId(), indexOK,
 										NULL, scanKeyCount, scanKey);
 
 	heapTuple = systable_getnext(scanDescriptor);
 	while (HeapTupleIsValid(heapTuple))
 	{
-		TupleDesc tupleDescriptor = RelationGetDescr(pgShardPlacement);
+		TupleDesc tupleDescriptor = RelationGetDescr(pgPlacement);
 
 		ShardPlacement *placement = TupleToShardPlacement(tupleDescriptor, heapTuple);
+
+		uint32 groupId = placement->groupId;
+		WorkerNode *worker = NodeForGroup(groupId);
+
+		if (worker == NULL)
+		{
+			ereport(ERROR, (errmsg("the metadata is inconsistent"),
```

Comment:
Writing a longer comment with examples but the short version is that leaving ShardPlacements with empty nodeName/nodePort is error-prone and requires a lot of changes. Error-prone because there's no indication that `nodeName` might be `NULL` and did you know `strncmp(str, NULL, n)` is undefined?

If being so trigger-happy about erroring out isn't okay and we want to change these call sites, I'd feel more comfortable if we made a `GroupShardPlacement` and a `NodeShardPlacement`, this function returns `GroupShardPlacements` and later functions (such as `FinalizedShardPlacementList`) resolve them to `NodeShardPlacements` when required.